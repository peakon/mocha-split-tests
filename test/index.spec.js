"use strict";

const {
  listFiles,
  loadRuntimeStats,
  addKnownRuntimes,
  distributeByRuntime
} = require("../lib/index");

describe("index", function() {
  describe("listFiles", function() {
    it("should fail when no files are found", function() {
      const pattern = "**/doesnotexist/*.js";
      expect(() => listFiles(pattern)).toThrow(
        `No files matching a pattern: ${pattern}`
      );
    });

    it("should return files when pattern is correct", function() {
      expect(listFiles("test/data/**/*.spec")).toMatchSnapshot();
    });
  });

  describe("loadRuntimeStats", function() {
    describe("when runtime.log contains malformed rows", function() {
      it("loads only well-formed rows from file", async function() {
        const stats = await loadRuntimeStats("./test/data/runtimes2.log");
        expect(stats.size).toEqual(3);
        expect(stats).toMatchSnapshot();
      });
    });

    describe("when runtime.log has no malformed rows", function() {
      it("loads all stats from a well-formed file", async function() {
        const stats = await loadRuntimeStats("./test/data/runtimes1.log");
        expect(stats.size).toEqual(11);
        expect(stats).toMatchSnapshot();
      });
    });

    describe("when runtime.log does not exist", function() {
      it("should fail with error", async function() {
        const stats = await loadRuntimeStats("/invalid/path/to/runtime.log");
        expect(stats.size).toEqual(0);
      });
    });
  });

  describe("addKnownRuntimes", function() {
    it("adds runtime stats to test backlog", function() {
      const testFiles = [
        { path: "test/1.spec.js", size: 1 },
        { path: "test/2.spec.js", size: 2 },
        { path: "test/3.spec.js", size: 3 }
      ];
      const runtimeStats = new Map()
        .set("test/1.spec.js", 100)
        .set("test/2.spec.js", 200)
        .set("test/4.spec.js", 400);
      expect(addKnownRuntimes(testFiles, runtimeStats)).toEqual([
        { path: "test/1.spec.js", size: 1, runtime: 100 },
        { path: "test/2.spec.js", size: 2, runtime: 200 },
        { path: "test/3.spec.js", size: 3, runtime: undefined }
      ]);
    });
  });

  describe("distributeByRuntime", function() {
    describe("when more than 90% of runtime stats are available", function() {
      it("distributes test based on runtime", async function() {
        const runtimeStats = new Map()
          .set("1", 100)
          .set("2", 200)
          .set("3", 300)
          .set("4", 400)
          .set("5", 500)
          .set("6", 600)
          .set("7", 700)
          .set("8", 800)
          .set("9", 900)
          .set("10", 1000);
        const testFiles = [
          { path: "1", size: 1 },
          { path: "2", size: 2 },
          { path: "3", size: 3 },
          { path: "4", size: 4 },
          { path: "5", size: 5 },
          { path: "6", size: 6 },
          { path: "7", size: 7 },
          { path: "8", size: 8 },
          { path: "9", size: 9 },
          { path: "10", size: 10 },
          { path: "11", size: 11 }
        ];
        const result = await distributeByRuntime({
          testFiles,
          runtimeStats,
          totalGroups: 4
        });
        expect(result).toMatchSnapshot();
      });
    });

    describe("when less than 90% of runtime stats are available", function() {
      it("distributes test files between buckets based on file size", async function() {
        const runtimeStats = new Map()
          .set("1", 100)
          .set("2", 200)
          .set("3", 300)
          .set("4", 400)
          .set("5", 500)
          .set("6", 600)
          .set("7", 700)
          .set("8", 800);
        const testFiles = [
          { path: "1", size: 1 },
          { path: "2", size: 2 },
          { path: "3", size: 3 },
          { path: "4", size: 4 },
          { path: "5", size: 5 },
          { path: "6", size: 6 },
          { path: "7", size: 7 },
          { path: "8", size: 8 },
          { path: "9", size: 9 }
        ];
        const result = await distributeByRuntime({
          testFiles,
          runtimeStats,
          totalGroups: 4
        });
        expect(result).toMatchSnapshot();
      });
    });
  });
});
