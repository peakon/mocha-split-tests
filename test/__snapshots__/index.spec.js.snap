// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index distributeByRuntime when less than 90% of runtime stats are available distributes test files between buckets based on file size 1`] = `
Array [
  Object {
    "files": Array [
      "9",
      "2",
      "1",
    ],
    "size": 12,
  },
  Object {
    "files": Array [
      "8",
      "3",
    ],
    "size": 11,
  },
  Object {
    "files": Array [
      "7",
      "4",
    ],
    "size": 11,
  },
  Object {
    "files": Array [
      "6",
      "5",
    ],
    "size": 11,
  },
]
`;

exports[`index distributeByRuntime when more than 90% of runtime stats are available distributes test based on runtime 1`] = `
Array [
  Object {
    "files": Array [
      "10",
      "3",
      "2",
    ],
    "size": 1500,
  },
  Object {
    "files": Array [
      "9",
      "4",
      "1",
    ],
    "size": 1400,
  },
  Object {
    "files": Array [
      "8",
      "5",
      "11",
    ],
    "size": 1300,
  },
  Object {
    "files": Array [
      "7",
      "6",
    ],
    "size": 1300,
  },
]
`;

exports[`index listFiles should return files when pattern is correct 1`] = `
Array [
  Object {
    "path": "test/data/fakeSpecs/2.spec",
    "size": 2,
  },
  Object {
    "path": "test/data/fakeSpecs/3.spec",
    "size": 3,
  },
  Object {
    "path": "test/data/fakeSpecs/4.spec",
    "size": 4,
  },
  Object {
    "path": "test/data/fakeSpecs/specs/1.spec",
    "size": 1,
  },
]
`;

exports[`index loadRuntimeStats when runtime.log contains malformed rows loads only well-formed rows from file 1`] = `
Map {
  "test/controllers/application_controller.spec.js" => 5084,
  "test/models/user.spec.js" => 5084,
  "test/does-not-exist-anymore.spec.js" => 234,
}
`;

exports[`index loadRuntimeStats when runtime.log has no malformed rows loads all stats from a well-formed file 1`] = `
Map {
  "test/controllers/application_controller.spec.js" => 5084,
  "test/models/user.spec.js" => 5084,
  "test/models/company.spec.js" => 1234,
  "test/routers/foo.spec.js" => 20,
  "test/config/config.spec.js" => 345,
  "test/models/account.spec.js" => 2435,
  "test/routers/bar.spec.js" => 20,
  "test/config/buz.spec.js" => 345,
  "test/models/aaa.spec.js" => 2435,
  "test/does-not-exist-anymore.spec.js" => 234,
  "test/does-not-exist-anymore_2.spec.js" => 234234,
}
`;
